interop-tests/src/lib.rs

```diff
@@ -0,0 +1,224 @@
+        boxed_transport,
+        Behaviour {
+            ping: ping::Behaviour::new(ping::Config::new().with_interval(Duration::from_secs(1))),
+            keep_alive: keep_alive::Behaviour,
+        },
+        local_peer_id,
+    )
+    .build();
+
+    log::info!("Running ping test: {}", swarm.local_peer_id());
+
+    // Run a ping interop test. Based on `is_dialer`, either dial the address
+    // retrieved via `listenAddr` key over the redis connection. Or wait to be pinged and have
+    // `dialerDone` key ready on the redis connection.
+    match is_dialer {
+        true => {
+            let result: Vec<String> = redis_client
+                .blpop("listenerAddr", test_timeout.as_secs())
+                .await?;
+            let other = result
+                .get(1)
+                .context("Failed to wait for listener to be ready")?;
+
+            let handshake_start = Instant::now();
+
+            swarm.dial(other.parse::<Multiaddr>()?)?;
+            log::info!("Test instance, dialing multiaddress on: {}.", other);
+
+            let rtt = loop {
+                if let Some(SwarmEvent::Behaviour(BehaviourEvent::Ping(ping::Event {
+                    result: Ok(rtt),
+                    ..
+                }))) = swarm.next().await
+                {
+                    log::info!("Ping successful: {rtt:?}");
+                    break rtt.as_millis() as f32;
+                }
+            };
+
+            let handshake_plus_ping = handshake_start.elapsed().as_millis() as f32;
+            Ok(format!(
+                r#"{{"handshakePlusOneRTTMillis": {handshake_plus_ping:.1}, "pingRTTMilllis": {rtt:.1}}}"#
+            ))
+        }
+        #[cfg(target_arch = "wasm32")]
+        false => bail!("Cannot run as listener in browser, please change is_dialer to true"),
+        #[cfg(not(target_arch = "wasm32"))]
+        false => {
+            log::info!(
+                "Test instance, listening for incoming connections on: {:?}.",
+                local_addr
+            );
+            let id = swarm.listen_on(local_addr.parse()?)?;
+
+            loop {
+                if let Some(SwarmEvent::NewListenAddr {
+                    listener_id,
+                    address,
+                }) = swarm.next().await
+                {
+                    if address.to_string().contains("127.0.0.1") {
+                        continue;
+                    }
+                    if listener_id == id {
+                        let ma = format!("{address}/p2p/{local_peer_id}");
+                        redis_client.rpush("listenerAddr", ma).await?;
+                        break;
+                    }
+                }
+            }
+
+            // Drive Swarm in the background while we await for `dialerDone` to be ready.
+            tokio::spawn(async move {
+                loop {
+                    swarm.next().await;
+                }
+            });
+            tokio::time::sleep(test_timeout).await;
```

Comment:
Additionally, we'd have to move the `Swarm::listen_on` into the `match` arm where `is_dialer` is `false`.

Reply:
So, please take a look at this commit https://github.com/libp2p/rust-libp2p/pull/4063/commits/0eeb12469b2b0b726c0bc40bb161d01e8c7ee91e
It looks like this is required to be started even if the node is a dialing node, because otherwise webrtc tests fails. Given that, I'm not sure if removing the `tokio` dependency is still helpful there, as we have to break from this logic in case of wasm anyway, it will no longer fail on `listen_on` because it won't trigger it. And I think this explicit bail is better than stubbing the `id` variable with `ListenerId::new()` without starting the listener

Reply:
Hmm, interesting. I think the problem here is that we always need a UDP socket, even if we are just the dialer. For QUIC, we implicitly create one just for that dial: https://github.com/libp2p/rust-libp2p/blob/2c6c058bb7b0d373e05835fead3f8e63e3240583/transports/quic/src/transport.rs#L201-L217

For WebRTC, we just fail: https://github.com/libp2p/rust-libp2p/blob/2c6c058bb7b0d373e05835fead3f8e63e3240583/transports/webrtc/src/tokio/transport.rs#L130-L136

I think it would be good to just fix the WebRTC implementation here because (a) it is a general short-coming and (b) should allow us to simplify the implementation in this PR.

Let me know if you are up for it, otherwise I'll do it later today.

Reply:
sorry, I missed this one. I may take a look into it but rather not yet today. Feel free to resolve it if you have some time for it